RTClib.cpp

```diff
@@ -361,6 +399,133 @@ char* DateTime::toString(char* buffer){
+            buffer[i] = '0' + ss / 10;
+            buffer[i + 1] = '0' + ss % 10;
+        }
+        if (buffer[i] == 'D' && buffer[i + 1] == 'D' && buffer[i + 2] == 'D') {
+            static PROGMEM const char day_names[] = "SunMonTueWedThuFriSat";
+            const char* p = &day_names[3 * dayOfTheWeek()];
+            buffer[i] = pgm_read_byte(p);
+            buffer[i + 1] = pgm_read_byte(p + 1);
+            buffer[i + 2] = pgm_read_byte(p + 2);
+        }
+        else
+            if (buffer[i] == 'D' && buffer[i + 1] == 'D') {
+                buffer[i] = '0' + d / 10;
+                buffer[i + 1] = '0' + d % 10;
+            }
+        if (buffer[i] == 'M' && buffer[i + 1] == 'M' && buffer[i + 2] == 'M') {
+            static PROGMEM const char month_names[] = "JanFebMarAprMayJunJulAugSepOctNovDec";
+            const char* p = &month_names[3 * (m - 1)];
+            buffer[i] = pgm_read_byte(p);
+            buffer[i + 1] = pgm_read_byte(p + 1);
+            buffer[i + 2] = pgm_read_byte(p + 2);
+        }
+        else
+            if (buffer[i] == 'M' && buffer[i + 1] == 'M') {
+                buffer[i] = '0' + m / 10;
+                buffer[i + 1] = '0' + m % 10;
+            }
+        if (buffer[i] == 'Y' && buffer[i + 1] == 'Y' && buffer[i + 2] == 'Y' && buffer[i + 3] == 'Y') {
+            buffer[i] = '2';
+            buffer[i + 1] = '0';
+            buffer[i + 2] = '0' + (yOff / 10) % 10;
+            buffer[i + 3] = '0' + yOff % 10;
+        }
+        else
+            if (buffer[i] == 'Y' && buffer[i + 1] == 'Y') {
+                buffer[i] = '0' + (yOff / 10) % 10;
+                buffer[i + 1] = '0' + yOff % 10;
+            }
+        if (buffer[i] == 'a' && buffer[i + 1] == 'a') {
+            if (isPM) {
+                buffer[i] = 'P';
+                buffer[i + 1] = 'M';
+            }
+            else {
+                buffer[i] = 'A';
+                buffer[i + 1] = 'M';
+            }
+        }
+    }
+    return buffer;
+}
+
+/**************************************************************************/
+/*!
+      @brief  Return hours reformatted into 12-Hour time
+      @return uint8_t hours
+*/
+/**************************************************************************/
+uint8_t DateTime::twelveHour() const {
+    if (hh == 0 || hh == 12) { //midnight or noon
+        return 12;
+    }
+    else
+        if (hh > 12) { //1 o'clock or later
+            return hh - 12;
+        }
+    else { //morning
+        return hh;
+    }
+}
+
+/**************************************************************************/
+/*!
+   @brief  Return whether the current time is PM
+   @return uint8_t False = AM, True = PM
+*/
+/**************************************************************************/
+uint8_t DateTime::isPM() const {
+    if (hh < 12) { //morning
+        return false;
+    }
+    else if (hh >= 12) { //evening
+        return true;
+    }
```

Comment:
```suggestion
    return hh >= 12;
```

Not only is this simpler, it should also make Travis happy.

Reply:
That, friend, is the beauty of open source! You can take my bulky code and make it super-efficient!

Now that we can do that function inline, should I move it to the .h file?

Also, could you please explain to me why a Travis build would fail? Thanks - I'm new to this!

Reply:
> Now that we can do that function inline, should I move it to the .h file?

I would think so. This .h file is already inlining quite a few trivial functions, so inlining this one would be consistent with the current style.

> Also, could you please explain to me why a Travis build would fail?

If you click on “Show all checks” next to “All checks have failed”, you will see a mention of Travis, with a “Details” link to the right. Follow the link, then the first link on the next page (“The build”). You will see the build output messages. Search for “error” and you will see:

```text
.../RTClib.cpp: In member function 'uint8_t DateTime::isPM() const':
.../RTClib.cpp:527:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1plus: some warnings being treated as errors
```

What happens is that the compiler saw this structure:

```c++
uint8_t DateTime::isPM() const {
    if (some_condition)
        return false;
    else if (some_other_condition)
        return true;
    // And what if both conditions are false?
    // What is the function then supposed to return?
}
```

Clearly the compiler is not being very clever, as both conditions cannot be simultaneously false. This error only happens with the ESP32 build chain. A better optimizing compiler would likely optimize away the second `if` test.

However, it never hurts to make the logic of the code obvious at the syntax level. If you write:

```c++
if (some_condition)
    return false;
else
    return true;
```

it is immediately obvious, even to the dumbest compiler, that only one condition has to be tested, and that a value is always returned. Some style guides even recommend to drop the `else` keyword in order to make the last `return` unconditional.

Reply:
> I would think so. This .h file is already inlining quite a few trivial functions, so inlining this one would be consistent with the current style.

My thoughts exactly.

Thanks for explaining Travis. I only made it about half-way through the link-chain! 😄